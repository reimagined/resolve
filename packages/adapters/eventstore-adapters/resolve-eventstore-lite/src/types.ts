export type AdapterPool = {
  config: {
    databaseFile: string
    secretsFile: string
    secretsTableName: string
    eventsTableName: string
    snapshotsTableName: string
  }
  secretsDatabase: any
  database: any
  eventsTableName: string
  snapshotsTableName: string
  secretsTableName: string
  escapeId: (source: string) => string
  escape: (source: string) => string
  memoryStore: any
}

export type AdapterSpecific = {
  sqlite: any
  tmp: any
  os: any
  fs: any
}
