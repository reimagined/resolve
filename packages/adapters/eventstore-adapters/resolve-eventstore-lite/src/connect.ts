import getLog from './js/get-log'
import connectEventStore from './js/connect'
import { AdapterPool, AdapterSpecific } from './types'

const SQLITE_BUSY = 'SQLITE_BUSY'
const randRange = (min: number, max: number): number =>
  Math.floor(Math.random() * (max - min + 1)) + min

const fullJitter = (retries: number): number =>
  randRange(0, Math.min(100, 2 * 2 ** retries))

const connectSecretsStore = async (
  pool: AdapterPool,
  specific: AdapterSpecific
): Promise<void> => {
  const log = getLog('connectSecretsStore')

  log.debug('connecting to secrets store database')

  const {
    escape,
    config: { secretsTableName = 'default', secretsFile = 'secrets.db' },
  } = pool

  log.verbose(`secretsTableName: ${secretsTableName}`)
  log.verbose(`secretsFile: ${secretsFile}`)

  for (let retry = 0; ; retry++) {
    try {
      const secretsDatabase = await specific.sqlite.open(secretsFile)

      log.debug('adjusting connection')
      await secretsDatabase.exec(`PRAGMA busy_timeout=1000000`)
      await secretsDatabase.exec(`PRAGMA encoding=${escape('UTF-8')}`)
      await secretsDatabase.exec(`PRAGMA synchronous=EXTRA`)
      await secretsDatabase.exec(`PRAGMA journal_mode=DELETE`)

      Object.assign(pool, {
        secretsDatabase,
        secretsTableName,
      })

      log.debug('secrets store database connected successfully')
      return
    } catch (error) {
      if (error && error.code === SQLITE_BUSY) {
        log.warn(`received SQLITE_BUSY error code, retrying`)
        await new Promise((resolve) => setTimeout(resolve, fullJitter(retry)))
      } else {
        log.error(error.message)
        log.verbose(error.stack)
        throw error
      }
    }
  }
}

const connect = async (
  pool: AdapterPool,
  specific: AdapterSpecific
): Promise<any> => {
  const log = getLog('connect')
  log.debug('connecting to sqlite databases')

  const escapeId = (str: string): string =>
    `"${String(str).replace(/(["])/gi, '$1$1')}"`
  const escape = (str: string): string =>
    `'${String(str).replace(/(['])/gi, '$1$1')}'`

  Object.assign(pool, {
    escapeId,
    escape,
  })

  await Promise.all([
    connectEventStore(pool, specific),
    connectSecretsStore(pool, specific),
  ])
  log.debug('connection to sqlite databases established')
}

export default connect
