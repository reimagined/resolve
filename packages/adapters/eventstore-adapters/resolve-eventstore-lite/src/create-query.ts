import { EventFilter } from 'resolve-eventstore-base'
import { AdapterPool } from './types'

const injectString = (pool: AdapterPool, value: string): string =>
  `${pool.escape(value)}`
const injectNumber = (pool: AdapterPool, value: number): string => `${+value}`

const createQuery = (
  pool: AdapterPool,
  { eventTypes, aggregateIds, startTime, finishTime }: EventFilter
): string => {
  const { escapeId } = pool

  const queryConditions = []
  if (eventTypes != null) {
    queryConditions.push(
      `${escapeId('type')} IN (${eventTypes
        .map(injectString.bind(null, pool))
        .join(', ')})`
    )
  }
  if (aggregateIds != null) {
    queryConditions.push(
      `${escapeId('aggregateId')} IN (${aggregateIds
        .map(injectString.bind(null, pool))
        .join(', ')})`
    )
  }
  if (startTime != null) {
    queryConditions.push(
      `${escapeId('timestamp')} > ${injectNumber(pool, startTime)}`
    )
  }
  if (finishTime != null) {
    queryConditions.push(
      `${escapeId('timestamp')} < ${injectNumber(pool, finishTime)}`
    )
  }

  return queryConditions.length > 0
    ? `WHERE ${queryConditions.join(' AND ')}`
    : ''
}

export default createQuery
