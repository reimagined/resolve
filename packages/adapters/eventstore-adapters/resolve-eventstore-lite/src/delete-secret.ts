import getLog from './get-log'
import { AdapterPool } from './types'

const deleteSecret = async (
  { database, secretsTableName, escapeId }: AdapterPool,
  selector: string
): Promise<void> => {
  const log = getLog('secretsManager:deleteSecret')
  log.debug(`removing secret from the database`)

  log.verbose(`selector: ${selector}`)
  log.verbose(`secretsTableName: ${secretsTableName}`)

  log.debug(`executing SQL query`)
  await database.exec(
    `DELETE FROM ${escapeId(secretsTableName)} WHERE id="${selector}"`
  )
  log.debug(`query executed successfully`)
}

export default deleteSecret
