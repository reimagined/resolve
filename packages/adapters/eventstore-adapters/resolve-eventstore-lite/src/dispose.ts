import getLog from './js/get-log'
import disposeEventStore from './js/dispose'
import { AdapterPool } from './types'

const disposeSecretsStore = (pool: AdapterPool): Promise<any> => {
  const { secretsDatabase } = pool
  return secretsDatabase.close()
}

const dispose = async (pool: AdapterPool): Promise<any> => {
  const log = getLog('dispose')

  log.debug(`disposing the event store`)
  await Promise.all([disposeEventStore(pool), disposeSecretsStore(pool)])
  log.debug(`the event store disposed`)
}

export default dispose
