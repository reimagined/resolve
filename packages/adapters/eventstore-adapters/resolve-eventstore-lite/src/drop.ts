import { EventstoreResourceNotExistError } from 'resolve-eventstore-base'
import getLog from './get-log'
import { AdapterPool } from './types'

const drop = async (pool: AdapterPool): Promise<any> => {
  const log = getLog('drop')
  log.debug(`dropping the event store`)

  const {
    database,
    eventsTableName,
    snapshotsTableName,
    secretsTableName,
    escapeId,
    memoryStore,
    config,
  } = pool

  const errors: any[] = []
  const statements: string[] = [
    `DROP TABLE IF EXISTS ${escapeId(`${eventsTableName}-freeze`)}`,
    `DROP TABLE ${escapeId(eventsTableName)}`,
    `DROP TABLE ${escapeId(snapshotsTableName)}`,
    `DROP TABLE ${escapeId(secretsTableName)}`,
  ]

  for (const statement of statements) {
    try {
      log.debug(`executing query`)
      log.verbose(statement)
      await database.exec(statement)
      log.debug(`query executed successfully`)
    } catch (error) {
      if (error != null) {
        let errorToThrow = error
        if (
          /^SQLITE_ERROR: no such (?:table|(?:Table|Index)).*?$/.test(
            error.message
          )
        ) {
          errorToThrow = new EventstoreResourceNotExistError(
            `sqlite adapter for database "${config.databaseFile}" already dropped`
          )
        } else {
          log.error(errorToThrow.message)
          log.verbose(errorToThrow.stack)
        }
        errors.push(errorToThrow)
      }
    } finally {
      if (memoryStore != null) {
        try {
          await memoryStore.drop()
        } catch (e) {
          log.error(e.message)
          log.verbose(e.stack)
          errors.push(e)
        }
      }
    }
  }

  pool.maybeThrowResourceError(errors)

  log.debug(`the event store dropped`)
}

export default drop
