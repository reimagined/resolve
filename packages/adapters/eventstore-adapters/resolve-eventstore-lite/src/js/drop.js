import getLog from './get-log'
import { EventstoreResourceNotExistError } from 'resolve-eventstore-base'

const dropEventStore = async ({
  database,
  eventsTableName,
  snapshotsTableName,
  escapeId,
  memoryStore,
  config,
}) => {
  const log = getLog('dropEventStore')
  try {
    log.debug(`dropping events freeze table`)
    await database.exec(
      `DROP TABLE IF EXISTS ${escapeId(`${eventsTableName}-freeze`)}`
    )

    log.debug(`dropping events primary table`)
    await database.exec(`DROP TABLE ${escapeId(eventsTableName)}`)

    log.debug(`dropping snapshots table`)
    await database.exec(`DROP TABLE ${escapeId(snapshotsTableName)}`)

    log.debug(`event store tables are dropped`)
  } catch (error) {
    if (error) {
      let errorToThrow = error
      if (/^SQLITE_ERROR: no such table.*?$/.test(error.message)) {
        errorToThrow = new EventstoreResourceNotExistError(
          `sqlite adapter for database "${config.databaseFile}" already dropped`
        )
      } else {
        log.error(errorToThrow.message)
        log.verbose(errorToThrow.stack)
      }
      throw errorToThrow
    }
  } finally {
    if (memoryStore != null) {
      try {
        await memoryStore.drop()
      } catch (e) {
        log.error(e.message)
        log.verbose(e.stack)
      }
    }
  }
}

export default dropEventStore
