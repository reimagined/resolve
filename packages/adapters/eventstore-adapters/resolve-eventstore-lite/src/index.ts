import createAdapter from 'resolve-eventstore-base'
import sqlite from 'sqlite'
import tmp from 'tmp'
import os from 'os'
import fs from 'fs'

import loadEventsByCursor from './js/load-events-by-cursor'
import loadEventsByTimestamp from './js/load-events-by-timestamp'
import getLatestEvent from './js/get-latest-event'
import saveEvent from './js/save-event'
import injectEvent from './js/inject-event'
import freeze from './js/freeze'
import unfreeze from './js/unfreeze'
import shapeEvent from './js/shape-event'
import loadSnapshot from './js/load-snapshot'
import dropSnapshot from './js/drop-snapshot'
import saveSnapshot from './js/save-snapshot'
import beginIncrementalImport from './js/begin-incremental-import'
import commitIncrementalImport from './js/commit-incremental-import'
import rollbackIncrementalImport from './js/rollback-incremental-import'
import pushIncrementalImport from './js/push-incremental-import'

import connect from './connect'
import init from './init'
import drop from './drop'
import dispose from './dispose'
import getSecretsManager from './secrets-manager'

const wrappedCreateAdapter = createAdapter.bind(null, {
  connect,
  loadEventsByCursor,
  loadEventsByTimestamp,
  getLatestEvent,
  saveEvent,
  init,
  drop,
  dispose,
  injectEvent,
  freeze,
  unfreeze,
  shapeEvent,
  getSecretsManager,
  saveSnapshot,
  dropSnapshot,
  loadSnapshot,
  beginIncrementalImport,
  commitIncrementalImport,
  rollbackIncrementalImport,
  pushIncrementalImport,
  sqlite,
  tmp,
  os,
  fs
})

export default wrappedCreateAdapter
