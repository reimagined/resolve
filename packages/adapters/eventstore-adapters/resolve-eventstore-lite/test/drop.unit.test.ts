/* eslint-disable import/no-extraneous-dependencies */
import { mocked } from 'ts-jest/utils'
/* eslint-enable import/no-extraneous-dependencies */

import { AdapterPool } from '../src/types'
import drop from '../src/drop'
import dropEventStore from '../src/js/drop'

jest.mock('../src/js/get-log')
jest.mock('../src/js/drop', () => jest.fn())

const mDropEventStore = mocked(dropEventStore)
const mExec = jest.fn()

let pool: AdapterPool

beforeEach(() => {
  pool = {
    config: {
      databaseFile: 'database-file',
      secretsFile: 'secret-file',
      secretsTableName: 'secrets-table',
      eventsTableName: 'table-name',
      snapshotsTableName: 'snapshots-table-name',
    },
    secretsDatabase: { exec: mExec },
    secretsTableName: 'secrets-table',
    database: '',
    eventsTableName: '',
    snapshotsTableName: '',
    escape: jest.fn(),
    escapeId: jest.fn((v: any) => `"${v}-escaped"`),
    memoryStore: 'memory',
  }
})

afterEach(() => {
  mDropEventStore.mockClear()
  mExec.mockClear()
})

test('event store dropped', async () => {
  await drop(pool)

  expect(mDropEventStore).toHaveBeenCalledWith(pool)
})

test('secrets store dropped', async () => {
  await drop(pool)

  expect(mExec.mock.calls).toMatchSnapshot('drop table with keys')
})
