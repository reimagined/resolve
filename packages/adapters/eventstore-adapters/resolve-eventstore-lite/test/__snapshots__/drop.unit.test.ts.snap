// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`secrets store dropped: drop table with keys 1`] = `
Array [
  Array [
    "DROP TABLE IF EXISTS ESCAPEID[table-name-freeze]",
  ],
  Array [
    "DROP TABLE ESCAPEID[table-name]",
  ],
  Array [
    "DROP TABLE ESCAPEID[snapshots-table-name]",
  ],
  Array [
    "DROP TABLE ESCAPEID[secrets-table]",
  ],
]
`;
